File: go/cameracontrol/client_test.go
package cameracontrol

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"errors"
	"testing"
	"time"
)

// fakeConn replays scripted server messages and answers requests the way the
// app does, so the handshake and event queue can be exercised offline.
type fakeConn struct {
	messages []Object
	sent     []Object
	closed   bool
}

func (c *fakeConn) Send(_ context.Context, message []byte) error {
	var payload Object
	if err := json.Unmarshal(message, &payload); err != nil {
		return err
	}
	c.sent = append(c.sent, payload)
	switch payload["action"] {
	case "authenticate":
		c.messages = append([]Object{{
			"id": payload["id"], "ok": true,
			"result": Object{"protocolVersion": ProtocolVersion},
		}}, c.messages...)
	case "status":
		c.messages = append(c.messages,
			Object{"event": "captureSession", "session": Object{"id": "session-1", "state": "waiting"}},
			Object{"id": payload["id"], "ok": true, "result": Object{"ready": true}},
		)
	}
	return nil
}

func (c *fakeConn) Receive(context.Context) ([]byte, error) {
	if len(c.messages) == 0 {
		return nil, errors.New("the connection was closed by the peer")
	}
	message := c.messages[0]
	c.messages = c.messages[1:]
	return json.Marshal(message)
}

func (c *fakeConn) Close() error {
	c.closed = true
	return nil
}

func newTestClient(t *testing.T, conn *fakeConn) *Client {
	t.Helper()
	client, err := New("wss://192.168.1.50:8080/ws", Options{
		Password:  "secret123",
		TLSConfig: &tls.Config{MinVersion: tls.VersionTLS12},
		Timeout:   time.Second,
		Dial: func(context.Context, string, *tls.Config) (Conn, error) {
			return conn, nil
		},
	})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return client
}

func TestAuthenticatesProtocolV4AndQueuesEvents(t *testing.T) {
	conn := &fakeConn{messages: []Object{{"event": "hello", "protocolVersion": ProtocolVersion}}}
	client := newTestClient(t, conn)
	ctx := context.Background()
	if err := client.Connect(ctx); err != nil {
		t.Fatalf("Connect: %v", err)
	}
	status, err := client.Status(ctx)
	if err != nil {
		t.Fatalf("Status: %v", err)
	}
	if status["ready"] != true {
		t.Fatalf("unexpected status: %v", status)
	}
	event, err := client.WaitForEvent(ctx, "captureSession", nil, 100*time.Millisecond)
	if err != nil {
		t.Fatalf("WaitForEvent: %v", err)
	}
	session := event["session"].(map[string]any)
	if session["id"] != "session-1" {
		t.Fatalf("unexpected event: %v", event)
	}
	if conn.sent[0]["action"] != "authenticate" {
		t.Fatalf("the first request must authenticate, got %v", conn.sent[0])
	}
	client.Close()
	if !conn.closed {
		t.Fatal("Close must close the socket")
	}
}

func TestRejectsAPIErrors(t *testing.T) {
	conn := &fakeConn{messages: []Object{
		{"event": "hello", "protocolVersion": ProtocolVersion},
		{"id": "go-2", "ok": false, "error": Object{"code": "camera_busy", "message": "Busy"}},
	}}
	client := newTestClient(t, conn)
	ctx := context.Background()
	if err := client.Connect(ctx); err != nil {
		t.Fatalf("Connect: %v", err)
	}
	_, err := client.Request(ctx, "cancelCaptureSession", Object{"sessionId": "session-1"}, false)
	var apiError *APIError
	if !errors.As(err, &apiError) {
		t.Fatalf("expected an APIError, got %v", err)
	}
	if apiError.Code != "camera_busy" {
		t.Fatalf("unexpected error code: %s", apiError.Code)
	}
}

func TestRejectsOldProtocolServers(t *testing.T) {
	conn := &fakeConn{messages: []Object{{"event": "hello", "protocolVersion": 1}}}
	client := newTestClient(t, conn)
	err := client.Connect(context.Background())
	var protocolError *ProtocolError
	if !errors.As(err, &protocolError) {
		t.Fatalf("expected a ProtocolError, got %v", err)
	}
}

func TestDecodePhotoRejectsForeignMedia(t *testing.T) {
	if _, err := DecodePhoto(Object{"mimeType": "image/png"}); err == nil {
		t.Fatal("a non-JPEG photo must be rejected")
	}
	jpeg, err := DecodePhoto(Object{"mimeType": "image/jpeg", "dataBase64": "/9j/4AAQ"})
	if err != nil {
		t.Fatalf("DecodePhoto: %v", err)
	}
	if len(jpeg) != 6 || jpeg[0] != 0xff || jpeg[1] != 0xd8 {
		t.Fatalf("unexpected JPEG bytes: %x", jpeg)
	}
}

func TestRequiresSecureURL(t *testing.T) {
	if _, err := New("ws://192.168.1.50:8080/ws", Options{Password: "x"}); err == nil {
		t.Fatal("a ws:// URL must be rejected")
	}
}

File: nodejs/test/client.test.js
import assert from 'node:assert/strict';
import test from 'node:test';

import {
  API_PROTOCOL_VERSION,
  ApiError,
  CameraControlClient,
  ProtocolError,
  decodePhoto,
} from '../src/index.js';

/**
 * Replays scripted server messages and answers requests the way the app does,
 * so the handshake and event queue can be exercised offline.
 */
class FakeChannel {
  constructor(messages) {
    this.messages = [...messages];
    this.sent = [];
    this.closed = false;
  }

  send(message) {
    const payload = JSON.parse(message);
    this.sent.push(payload);
    if (payload.action === 'authenticate') {
      this.messages.unshift({
        id: payload.id,
        ok: true,
        result: { protocolVersion: API_PROTOCOL_VERSION },
      });
    } else if (payload.action === 'status') {
      this.messages.push(
        { event: 'captureSession', session: { id: 'session-1', state: 'waiting' } },
        { id: payload.id, ok: true, result: { ready: true } },
      );
    }
  }

  async receive() {
    if (this.messages.length === 0) throw new Error('the connection was closed by the peer');
    return JSON.stringify(this.messages.shift());
  }

  close() {
    this.closed = true;
  }
}

function makeClient(channel) {
  return new CameraControlClient('wss://192.168.1.50:8080/ws', 'secret123', undefined, {
    timeout: 1_000,
    connectFactory: async () => channel,
  });
}

test('authenticates protocol v4 and queues events', async () => {
  const channel = new FakeChannel([{ event: 'hello', protocolVersion: API_PROTOCOL_VERSION }]);
  const client = makeClient(channel);
  await client.connect();
  assert.deepEqual(await client.status(), { ready: true });
  const event = await client.waitForEvent('captureSession', { timeout: 100 });
  assert.equal(event.session.id, 'session-1');
  assert.equal(channel.sent[0].action, 'authenticate');
  client.close();
  assert.equal(channel.closed, true);
});

test('rejects API errors', async () => {
  const channel = new FakeChannel([
    { event: 'hello', protocolVersion: API_PROTOCOL_VERSION },
    { id: 'node-2', ok: false, error: { code: 'camera_busy', message: 'Busy' } },
  ]);
  const client = makeClient(channel);
  await client.connect();
  await assert.rejects(
    client.request('cancelCaptureSession', { sessionId: 'session-1' }, { retryOnDisconnect: false }),
    (error) => error instanceof ApiError && error.code === 'camera_busy',
  );
});

test('rejects old protocol servers', async () => {
  const channel = new FakeChannel([{ event: 'hello', protocolVersion: 1 }]);
  const client = makeClient(channel);
  await assert.rejects(client.connect(), ProtocolError);
});

test('requires a wss URL', () => {
  assert.throws(
    () => new CameraControlClient('ws://192.168.1.50:8080/ws', 'x', undefined, {}),
    /wss:\/\/ URL/,
  );
});

test('decodes only JPEG photos', () => {
  assert.throws(() => decodePhoto({ mimeType: 'image/png' }), ProtocolError);
  const jpeg = decodePhoto({ mimeType: 'image/jpeg', dataBase64: '/9j/4AAQ' });
  assert.equal(jpeg[0], 0xff);
  assert.equal(jpeg[1], 0xd8);
});
